File: src/main/java/com/github/squirrelgrip/meter/service/Line500Processor.java
package com.github.squirrelgrip.meter.service;

import com.github.squirrelgrip.meter.domain.ContextKey;
import com.github.squirrelgrip.meter.domain.Version;
import com.github.squirrelgrip.meter.exception.InvalidVersionException;
import com.github.squirrelgrip.meter.exception.NMINotSetException;
import com.github.squirrelgrip.meter.exception.VersionNotSetException;

import java.util.Map;

public class Line500Processor implements LineProcessor {
    @Override
    public void preProcess(Map<ContextKey, Object> context) {
        if (!context.containsKey(ContextKey.VERSION)) {
            throw new VersionNotSetException(context);
        }
        if (context.get(ContextKey.VERSION) != Version.NEM12) {
            throw new InvalidVersionException("500", context);
        }
        if (!context.containsKey(ContextKey.NMI)) {
            throw new NMINotSetException(context);
        }
    }

    @Override
    public void process(String[] tokens, Map<ContextKey, Object> context) {
        context.remove(ContextKey.NMI);
        context.remove(ContextKey.INTERVAL_LENGTH);
    }
}

File: src/main/kotlin/com/github/squirrelgrip/meter/service/Line500Processor.kt
package com.github.squirrelgrip.meter.service

import com.github.squirrelgrip.meter.domain.ContextKey
import com.github.squirrelgrip.meter.domain.MeterContext
import com.github.squirrelgrip.meter.domain.Version

class Line500Processor(): BaseLineProcessor() {
    override fun process(
        context: MeterContext,
        tokens: Array<String>
    ) {
        context.requiresVersion("500", Version.NEM12)
        context.requiresNmi()

        context.remove(ContextKey.NMI)
        context.remove(ContextKey.INTERVAL_LENGTH)
    }
}
